### DISCUSSION TITLE
Aranet CO2 monitor

### DISCUSSION THREAD

**Discussion by Opiekelly** [theengs/gateway#189] [Ideas]
Hello I have a couple of Aranet4 CO2 monitors.  Would be willing to capture data/debug to support this device.

Thanks
Brendan

**Comment by deepcoder** [comment]
Here is some code for the gateway that might help you move forward. I apologize, I got side tracked and never had a chance to help put this properly into theengs decoder or esp code.

https://gist.github.com/deepcoder/3aa75b2368e048c030bc5f79f9870680

I've very happy with these Aranet devices, bit pricey, however seem very accurate.

Good hunting!

**Reply by DigiH** [reply]
I remember commenting your gist, and things do look nice and clear there. All we'd really need is some sample broadcast advertising data to create and verify a decoder for the Aranet CO2 monitor.

> Would be willing to capture data/debug to support this device.

If you could set 

```
-padv PUBLISH_ADVDATA, --publish_advdata PUBLISH_ADVDATA
                        Publish advertising and advanced data (1) or not (0)
                        (default: 0)
```

to **1** at starting Theengs Gateway and paste some sample messages from your Aranet4 CO2 monitors with the undecoded raw data here.

**Reply by Opiekelly** [reply]
I could not get the sample code from gist to subscribe correctly and print data.

`(aranet4) ➜  aranet4 git:(main) ✗ ./aranet4.py
Program start : aranet4-decode Version : 1.6`

Added print statement:
`def on_message(mqttc, userdata, message):
    msg = json.loads(str(message.payload.decode("utf-8")))
    print(msg)
`
update subscriber:
` #mqttc.subscribe("home/OpenMQTTGateway/BTtoMQTT/#")
    mqttc.subscribe("home/+/BTtoMQTT/#")`

wireshark packet captures if that helps:
`MQ Telemetry Transport Protocol, Publish Message
    [Expert Info (Note/Protocol): Unknown version (missing the CONNECT packet?)]
    Header Flags: 0x30, Message Type: Publish Message, QoS Level: At most once delivery (Fire and Forget)
        0011 .... = Message Type: Publish Message (3)
        .... 0... = DUP Flag: Not set
        .... .00. = QoS Level: At most once delivery (Fire and Forget) (0)
        .... ...0 = Retain: Not set
    Msg Len: 183
    Topic Length: 41
    Topic: home/TheengsGateway/BTtoMQTT/7E4853FEBCEA
    Message: 7b226e616d65223a2022426173656d656e7420737065616b6572222c20226964223a2022…

MQ Telemetry Transport Protocol, Publish Message
    [Expert Info (Note/Protocol): Unknown version (missing the CONNECT packet?)]
    Header Flags: 0x30, Message Type: Publish Message, QoS Level: At most once delivery (Fire and Forget)
        0011 .... = Message Type: Publish Message (3)
        .... 0... = DUP Flag: Not set
        .... .00. = QoS Level: At most once delivery (Fire and Forget) (0)
        .... ...0 = Retain: Not set
    Msg Len: 186
    Topic Length: 41
    Topic: home/TheengsGateway/BTtoMQTT/4831B7808546
    Message: 7b226d616e75666163747572657264617461223a20226430303630313436383538306237…

MQ Telemetry Transport Protocol, Publish Message
    [Expert Info (Note/Protocol): Unknown version (missing the CONNECT packet?)]
    Header Flags: 0x30, Message Type: Publish Message, QoS Level: At most once delivery (Fire and Forget)
        0011 .... = Message Type: Publish Message (3)
        .... 0... = DUP Flag: Not set
        .... .00. = QoS Level: At most once delivery (Fire and Forget) (0)
        .... ...0 = Retain: Not set
    Msg Len: 186
    Topic Length: 41
    Topic: home/TheengsGateway/BTtoMQTT/4831B7808546
    Message: 7b226d616e75666163747572657264617461223a20226430303630313436383538306237…

MQ Telemetry Transport Protocol, Publish Message
    [Expert Info (Note/Protocol): Unknown version (missing the CONNECT packet?)]
    Header Flags: 0x30, Message Type: Publish Message, QoS Level: At most once delivery (Fire and Forget)
        0011 .... = Message Type: Publish Message (3)
        .... 0... = DUP Flag: Not set
        .... .00. = QoS Level: At most once delivery (Fire and Forget) (0)
        .... ...0 = Retain: Not set
    Msg Len: 224
    Topic Length: 41
    Topic: home/TheengsGateway/BTtoMQTT/C4D4F6CBE513
    Message: 7b226d616e75666163747572657264617461223a20223032303732313065303430313030…

MQ Telemetry Transport Protocol, Publish Message
    [Expert Info (Note/Protocol): Unknown version (missing the CONNECT packet?)]
    Header Flags: 0x30, Message Type: Publish Message, QoS Level: At most once delivery (Fire and Forget)
        0011 .... = Message Type: Publish Message (3)
        .... 0... = DUP Flag: Not set
        .... .00. = QoS Level: At most once delivery (Fire and Forget) (0)
        .... ...0 = Retain: Not set
    Msg Len: 183
    Topic Length: 41
    Topic: home/TheengsGateway/BTtoMQTT/7E4853FEBCEA
    Message: 7b226e616d65223a2022426173656d656e7420737065616b6572222c20226964223a2022…

MQ Telemetry Transport Protocol, Publish Message
    [Expert Info (Note/Protocol): Unknown version (missing the CONNECT packet?)]
    Header Flags: 0x30, Message Type: Publish Message, QoS Level: At most once delivery (Fire and Forget)
        0011 .... = Message Type: Publish Message (3)
        .... 0... = DUP Flag: Not set
        .... .00. = QoS Level: At most once delivery (Fire and Forget) (0)
        .... ...0 = Retain: Not set
    Msg Len: 224
    Topic Length: 41
    Topic: home/TheengsGateway/BTtoMQTT/C4D4F6CBE513
    Message: 7b226d616e75666163747572657264617461223a20223032303732313065303430313030…

MQ Telemetry Transport Protocol, Publish Message
    [Expert Info (Note/Protocol): Unknown version (missing the CONNECT packet?)]
    Header Flags: 0x30, Message Type: Publish Message, QoS Level: At most once delivery (Fire and Forget)
        0011 .... = Message Type: Publish Message (3)
        .... 0... = DUP Flag: Not set
        .... .00. = QoS Level: At most once delivery (Fire and Forget) (0)
        .... ...0 = Retain: Not set
    Msg Len: 186
    Topic Length: 41
    Topic: home/TheengsGateway/BTtoMQTT/4831B7808546
    Message: 7b226d616e75666163747572657264617461223a20226430303630313436383538306237…

MQ Telemetry Transport Protocol, Publish Message
    [Expert Info (Note/Protocol): Unknown version (missing the CONNECT packet?)]
    Header Flags: 0x30, Message Type: Publish Message, QoS Level: At most once delivery (Fire and Forget)
        0011 .... = Message Type: Publish Message (3)
        .... 0... = DUP Flag: Not set
        .... .00. = QoS Level: At most once delivery (Fire and Forget) (0)
        .... ...0 = Retain: Not set
    Msg Len: 183
    Topic Length: 41
    Topic: home/TheengsGateway/BTtoMQTT/7E4853FEBCEA
    Message: 7b226e616d65223a2022426173656d656e7420737065616b6572222c20226964223a2022…

MQ Telemetry Transport Protocol, Publish Message
    [Expert Info (Note/Protocol): Unknown version (missing the CONNECT packet?)]
    Header Flags: 0x30, Message Type: Publish Message, QoS Level: At most once delivery (Fire and Forget)
        0011 .... = Message Type: Publish Message (3)
        .... 0... = DUP Flag: Not set
        .... .00. = QoS Level: At most once delivery (Fire and Forget) (0)
        .... ...0 = Retain: Not set
    Msg Len: 224
    Topic Length: 41
    Topic: home/TheengsGateway/BTtoMQTT/C4D4F6CBE513
    Message: 7b226d616e75666163747572657264617461223a20223032303732313065303430313030…

`

**Reply by Opiekelly** [reply]
Ok I set username and paasword and have this print message.

`{'manufacturerdata': '0207210e0401000c0f016f030802681f1362012c018c00b9', 'name': 'Aranet4 221CB', 'id': 'C4:D4:F6:CB:E5:13', 'rssi': -77, 'mfr': 'Akciju sabiedriba "SAF TEHNIKA"'}
C4:D4:F6:CB:E5:13
received message:  {'manufacturerdata': '0207210e0401000c0f016f030802681f1362012c018c00b9', 'name': 'Aranet4 221CB', 'id': 'C4:D4:F6:CB:E5:13', 'rssi': -77, 'mfr': 'Akciju sabiedriba "SAF TEHNIKA"'}
20231219043450 C4:D4:F6:CB:E5:13 00000010 1 4 0 0`

**Reply by DigiH** [reply]
Were your values for the above raw data these?

`"tempc":26,"tempf":78.8,"hum":19,"pres":804,"co2":879,"batt":98`

They look all fine to me, apart from the very dry "hum":19 ;)

If you could give another manufacturerdata sample, this time along with the values showing on the respective Aranet4. Then we should be ready to include an Aranet4 decoder into Theengs Decoder/Gateway.

**Reply by Opiekelly** [reply]
Here you go.
`{'manufacturerdata': '0207210e0401000c0f014c030e02601f1162012c01b10036', 'name': 'Aranet4 221CB', 'id': 'C4:D4:F6:CB:E5:13', 'rssi': -30, 'mfr': 'Akciju sabiedriba "SAF TEHNIKA"'}
C4:D4:F6:CB:E5:13
received message:  {'manufacturerdata': '0207210e0401000c0f014c030e02601f1162012c01b10036', 'name': 'Aranet4 221CB', 'id': 'C4:D4:F6:CB:E5:13', 'rssi': -30, 'mfr': 'Akciju sabiedriba "SAF TEHNIKA"'}
20231219150015 C4:D4:F6:CB:E5:13 00000010 1 4 0 0`

<img width="403" alt="Screenshot 2023-12-19 at 8 00 50 AM" src="https://github.com/theengs/gateway/assets/760918/3f7f3001-7897-4b58-8ee6-ed9cff715d0e">

**Reply by DigiH** [reply]
Perfect., thanks!

All clear here, with the new Aranet4 decoder resolving the above into

`{"brand":"Aranet","model":"Aranet4 CO₂ Monitor","model_id":"ARANET4","type":"AIR","acts":true,"tempc":26.3,"tempf":79.34,"hum":17,"pres":803.2,"co2":844,"batt":98}`

only with the difference that we do publish the air pressure in hectopascal (hPa) instead of your psi, but that should/could be converted in HA AFAIK. And the previous low humidity also was correct :) 

Just one more questions to finalise the Aranet4 decoder:
• If you set Theengs Gateway to passive scanning only, does it still get and publish the same  messages with the full manufacturerdata, or does the device require active scanning?

> 
> -s {active,passive}, --scanning_mode {active,passive}
>                         Scanning mode (default: active)

I assume so (hence I added the `"acts":true`), but would like to make sure to be able to allow for the correct automatic passive/active scanning definition.

**Reply by Opiekelly** [reply]
Yes this works.  I am in Colorado so humidity is low.  I use the CO2 monitor to make sure the ppm is low for gourmet mushroom growing.  The monitor is just sitting on my desk for testing so the humidity is low as well as CO2.
`(test) ➜  aranet4 git:(main) ✗ ./aranet4.py
Program start : aranet4-decode Version : 1.6
{'manufacturerdata': '0207210e0401000c0f013103f601671f1262012c01540057', 'name': 'Aranet4 221CB', 'id': 'C4:D4:F6:CB:E5:13', 'rssi': -27, 'mfr': 'Akciju sabiedriba "SAF TEHNIKA"'}
C4:D4:F6:CB:E5:13
received message:  {'manufacturerdata': '0207210e0401000c0f013103f601671f1262012c01540057', 'name': 'Aranet4 221CB', 'id': 'C4:D4:F6:CB:E5:13', 'rssi': -27, 'mfr': 'Akciju sabiedriba "SAF TEHNIKA"'}
20231219174339 C4:D4:F6:CB:E5:13 00000010 1 4 0 0
`

**Reply by Opiekelly** [reply]
When you check in the code can I get access to the repo so I can add to home assistant and test?

**Reply by DigiH** [reply]
> • If you set Theengs Gateway to passive scanning only, does it still get and publish the same messages with the full manufacturerdata, or does the device require active scanning?

> Yes this works.

> ./aranet4.py Program start : aranet4-decode Version : 1.6 {'manufacturerdata': '0207210e0401000c0f013103f601671f1262012c01540057', 'name': 'Aranet4 221CB', 'id': 'C4:D4:F6:CB:E5:13', 'rssi': -27, 'mfr': 'Akciju sabiedriba "SAF TEHNIKA"'}

I'm afraid this was with the aranet4.py integration again, which seem to do an active scan by default - the broadcast name is a giveaway ;)

Any chance you could try with Theengs Gateway and setting it to passive scanning as described and liked to above? If you also set 

> -padv PUBLISH_ADVDATA, --publish_advdata PUBLISH_ADVDATA
>                         Publish advertising and advanced data (1) or not (0)
>                         (default: 0)

to **1** you should see if the manufacturerdata is present with passive scanning or not. Best to monitor it with [MQTT Explorer](https://github.com/thomasnordquist/MQTT-Explorer/releases/tag/v0.3.5).

In general passive scanning has not effect on the battery of the sensor, while active scanning does, and therefore affects the battery life. So passive scanning is always preferable when a device allows for it.

> When you check in the code can I get access to the repo so I can add to home assistant and test?

Sure, depending on how you installed Theengs Gateway. 
If you did a pip install you can just follow the manual [build and install of the latest development state](https://decoder.theengs.io/use/python.html#installing-a-development-version) of [Theengs Decoder](https://github.com/theengs/decoder) and you will get the Aranet4 decoder after it has been merged.

If you use the Docker image, Snap or Theengs Gateway HA Add-on it would be a bit more involved, but as all projects are open source here on GitHub it would also be possible.

Alternatively you would have to wait for the next release of Theengs Decoder and the subsequent Theengs Gateway release.

**Reply by Opiekelly** [reply]
So the aranet4.py code is a subscriber to the MQTT broker.  It is not reading from the hci0 interface.  I have a system I call ha2 with theengs gateway as an add on and it publishes to ha1.  On ha1 I have an mqtt broker and ha2 subscribes to those events via aranet4.py.  All of my environment is in HA and I don't use other systems with docker or linux.


<img width="1243" alt="Screenshot 2023-12-19 at 12 20 53 PM" src="https://github.com/theengs/gateway/assets/760918/33faf548-efd6-443f-8cf4-5059987063b1">
Screenshot showing TheengsGateway config on ha2.

mqtt explorer output:
<img width="1398" alt="Screenshot 2023-12-19 at 12 27 48 PM" src="https://github.com/theengs/gateway/assets/760918/2e79988c-539e-4e2a-bf30-aa6f6f271415">

**Reply by DigiH** [reply]
Thanks, the Theengs Gateway Add-On does show passive scanning in its Configuration. I am just confused to still see `'name': 'Aranet4 221CB'` in the messages, as the name usually is only broadcast when active scanning happens.

So I am taking out the active scanning requirement and am submitting the decoder now …

**Reply by DigiH** [reply]
Merged into Theengs Decoder

https://github.com/theengs/decoder

with the latest commit.

**Reply by DigiH** [reply]
I'll be issuing a new release of Theengs Decoder before I go away for the holidays on the weekend.

If you don't want to wait until the next official release of the Theengs Gateway Add-On, after that you can just uninstall the Add-On and reinstall it once the new Theengs Decoder version is out. 
The Add-On will then pull in the latest new Theengs Decoder version with the Aranet4 decoder included.

**Reply by Opiekelly** [reply]
Sounds good.  Thanks for getting this done!  Happy Holidays.

**Reply by DigiH** [reply]
Thanks for your collaboration, and Happy Holidays to you, too!

**Reply by Opiekelly** [reply]
Doesn't look like this will be picked up with a uninstall and install as the DockerFile is pointing to the gateway code.  Looks like I will have to wait for the next gateway code release.

Thanks

`ARG BUILD_FROM
FROM $BUILD_FROM
LABEL io.hass.version="1.9.0" io.hass.type="addon" io.hass.arch="armv7|armhf|aarch64|i386|amd64"
RUN apt-get update && apt-get upgrade -y && \
    apt-get install --no-install-recommends -y build-essential bluez python3-pip python3-dev && \
    pip3 install --extra-index-url=https://www.piwheels.org/simple TheengsGateway==1.2.0
ENV current /root
ENV HOME $current
WORKDIR $current
COPY run.sh /root
COPY entrypoint.sh /root
RUN chmod a+x /root/run.sh
RUN chmod a+x /root/entrypoint.sh
ENTRYPOINT ["/root/entrypoint.sh"]
CMD [ "/root/run.sh" ]`

**Comment by 1technophile** [comment]
@DigiH may be interested by this
